# ***Advent of Code 2024 - Day 04***

## ***Part A***

***Observation***

This problem is more like graph traversal. My approach would be:
1. First, we'd need to read and parse the input into a 2D grid
2. From each position containing 'X', we'd need to check all 8 possible directions
3. For each direction, we'd need to check if the next 3 characters spell "MAS"

In [14]:
# Read the input.txt file and parse into a list of lists
# - Open & read the file
# - Split content into row
# - Convert each row into list of characters
with open('input.txt', 'r') as file:
    content = file.read()
print(content[:100])

char_grid = content
char_grid = [list(line.strip()) for line in content.split('\n') if line]
print(f"Number of rows: {len(char_grid)}")
print(f"Number of columns: {len(char_grid[0])} in each row")

SXXSAMSSMMSMMSXXXXMMXMMMMSMMMSSSXSAMXMXSXMASXMSMXSMMXSMSXMMASXMASMSSMMXSMSSSSSSXSXMAMXAMXSMSXSXXMMMM
Number of rows: 140
Number of columns: 140 in each row


***Observation***:

Now we got a 140*140 grid of character, easier to traverse & access to each character. Move on to next 02 stages:
2. From each position containing 'X', we'd need to check all 8 possible directions
3. For each direction, we'd need to check if the next 3 characters spell "MAS"

We'll break it down into following steps:
 - Find all X positions
 - Define the eight possible direction: horizontal, vertical, diagonal
 - Check each direction of each X, validate if potential XMAS pattern met.

In [15]:
# Find all X positions
def find_x_positions(grid):
    x_positions = []
    for row_index in range(0,len(char_grid)):
        for col_index in range(0, len(char_grid[0])):
            if char_grid[row_index][col_index] == "X": x_positions.append((row_index, col_index))
    return x_positions

# Define the eight possible directions
directions = [
    (0, 1),   # right
    (1, 1),   # right-down diagonal
    (1, 0),   # down
    (1, -1),  # left-down diagonal
    (0, -1),  # left
    (-1, -1), # left-up diagonal
    (-1, 0),  # up
    (-1, 1)   # right-up diagonal
]

# Check whether direction met MAS patther
def check_for_XMAS_SAMX(grid, x_row, x_col, direction):
    rows, cols = len(grid), len(grid[0])
    
    # Calculate all four positions
    positions = [
        (x_row, x_col),  # X position
        (x_row + direction[0], x_col + direction[1]),
        (x_row + 2*direction[0], x_col + 2*direction[1]),
        (x_row + 3*direction[0], x_col + 3*direction[1])
    ]
    
    # Boundary check
    if any(x < 0 or x >= rows or y < 0 or y >= cols for x, y in positions):
        return 0
    
    # Get the sequence
    letters = ''.join(grid[x][y] for x, y in positions)
    
    # Count matches for both patterns
    matches = 0
    if letters == "XMAS":
        matches += 1
    if letters == "SAMX":
        matches += 1
    
    return matches


In [16]:
# Count XMAS pattern
def count_XMAS_pattern(grid, directions):
    count = 0
    x_positions = find_x_positions(grid)
    for position in x_positions:
        for direction in directions:
            count += check_for_XMAS_SAMX(grid, position[0], position[1], direction)
    return count

count_XMAS_pattern(char_grid, directions)

2593

In [17]:
class AdventDay4:

    def __init__(self):
        """Initialize the word search, the row and column dimensions, and
        the goals for the word search. The word search is stored as a list of
        lists where each list is a row of the word search. The goals are the
        words 'XMAS' and 'MAS' and the words 'SAMX' and 'SAM' for part 1 and
        part 2 respectively.
        """
        self.ws = []
        with open('input.txt', 'r') as f:
            for line in f.readlines():
                self.ws.append([x for x in line if x != '\n'])
        self.n = len(self.ws)
        self.m = len(self.ws[0])
        self.word_goal = ['XMAS', 'SAMX']
        self.x_goal = ['MAS', 'SAM']
        self.part1 = 0
        self.part2 = 0

    def xmas_search(self):
        """Search for the word 'XMAS' in the word search."""
        self._row_search()
        self._col_search()
        self._diag_search()

    def _row_search(self):
        """Search for the word 'XMAS' in the rows of the word search."""
        for row in self.ws:
            for i in range(self.m - 3):
                self.part1 += ''.join(row[i:i+4]) in self.word_goal

    def _col_search(self):
        """Search for the word 'XMAS' in the columns of the word search."""
        for col in zip(*self.ws):
            for i in range(self.m - 3):
                self.part1 += ''.join(col[i:i+4]) in self.word_goal

    def _diag_search(self):
        """Search for the word 'XMAS' in the diagonals of the word search."""
        for row in range(self.n - 3):
            for col in range(self.m - 3):
                diag_1 = ''.join([self.ws[row][col],
                                  self.ws[row+1][col+1],
                                  self.ws[row+2][col+2],
                                  self.ws[row+3][col+3]])
                diag_2 = ''.join([self.ws[row+3][col],
                                  self.ws[row+2][col+1],
                                  self.ws[row+1][col+2],
                                  self.ws[row][col+3]])
                self.part1 += diag_1 in self.word_goal
                self.part1 += diag_2 in self.word_goal

    def diag_x_mas_search(self):
        """Search for the word 'MAS' to cross in the diagonals of the word
        search.
        """
        for row in range(self.n - 2):
            for col in range(self.m - 2):
                diag_1 = ''.join([self.ws[row][col],
                                  self.ws[row+1][col+1],
                                  self.ws[row+2][col+2]])
                diag_2 = ''.join([self.ws[row+2][col],
                                  self.ws[row+1][col+1],
                                  self.ws[row][col+2]])
                self.part2 += diag_1 in self.x_goal and diag_2 in self.x_goal


if __name__ == '__main__':
    day4 = AdventDay4()
    day4.xmas_search()
    day4.diag_x_mas_search()
    print(day4.part1, day4.part2)

2593 1950


In [18]:
def count_XMAS_patterns(grid):
    rows = len(grid)
    cols = len(grid[0])
    total_count = 0
    
    # 8 directions: right, down-right, down, down-left, left, up-left, up, up-right
    directions = [
        (0,1), (1,1), (1,0), (1,-1), 
        (0,-1), (-1,-1), (-1,0), (-1,1)
    ]
    
    def check_pattern(x, y, dx, dy):
        positions = [
            (x, y),
            (x + dx, y + dy),
            (x + 2*dx, y + 2*dy),
            (x + 3*dx, y + 3*dy)
        ]
        
        # Boundary check
        if any(r < 0 or r >= rows or c < 0 or c >= cols for r,c in positions):
            return 0
            
        # Get sequence
        word = ''.join(grid[r][c] for r,c in positions)
        return (word == "XMAS") + (word == "SAMX")
    
    # Search grid
    for row in range(rows):
        for col in range(cols):
            for dx, dy in directions:
                total_count += check_pattern(row, col, dx, dy)
                
    return total_count

count_XMAS_patterns(char_grid)

5186